pkg/validation/apiextensions/v1/composition/custom_validator.go

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright 2023 the Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package composition
+
+import (
+	"context"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	xperrors "github.com/crossplane/crossplane-runtime/pkg/errors"
+	v1 "github.com/crossplane/crossplane/apis/apiextensions/v1"
+)
+
+// CustomValidator gathers required information using the provided client.Reader and then use them to render and
+// validated a Composition.
+type CustomValidator struct{}
```

Comment:
We expect this to be directly used by other users, e.g. a linter, right now it's only validating composition running "logically", but later on it will also validate them using the provided schemas.
Maybe I'd add back the [CustomValidator](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/webhook/admission/validator_custom.go#L31) implementation we had in the beginning as we are using directly the handler only while waiting for https://github.com/kubernetes-sigs/controller-runtime/issues/1896, so that we don't have to change the API later on. @lsviben WDYT?

Reply:
Discussed it with @lsviben, we agreed to move the handler to an internal package as you were suggesting and did some more refactoring along the way, see https://github.com/crossplane/crossplane/pull/3921/commits/916a6537996ad5dc8178f97a4781c8feab439e34.

CustomValidator has been renamed Validator, as it's not anymore implementing the CustomValidator interface and we kept it publicly available for the reasons explained above.

FYI: I would have loved to implement the CustomValidator interface on the Composition struct itself, as it would have been much more idiomatic, but unfortunately for the future development we have in mind we'll need to have access to other packages that rely on `github.com/crossplane/crossplane/apis/apiextensions/v1` itself and doing so would result in a dependency cycle. So we decided to keep "logical validation" code near to the struct and "more complex validation" code in `pkg/validation/...`.